Change `dim` to `axis` for tf.nn.softmax and tf.nn.log_softmax

yongtang opened the pull request:
This fix tries to address the issue raised in #7391 where `dim` was not changed to `axis` for tf.nn.softmax and tf.nn.log_softmax.

This is inconsistent with other ops in tensorflow.

This fix adds axis while at the same time keeps dim so that backward compatibility is maintained.

This fix fixes #7391.

This fix fixes #14191.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Change `dim` to `axis` for tf.nn.softmax and tf.nn.log_softmax

This fix tries to address the issue raised in 7391 where
`dim` was not changed to `axis` for tf.nn.softmax and tf.nn.log_softmax.
This is inconsistent with other ops in tensorflow.

This fix adds axis while at the same time keeps dim so that
backward compatibility is maintained.

This fix fixes 7391.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

tensorflow-jenkins commented:
Can one of the admins verify this patch?

martinwicke commented on tensorflow/python/ops/nn_ops.py:
```diff
@@ -1460,51 +1460,65 @@ def _swap_axis(logits, dim_index, last_index, name=None):
 
 
-def softmax(logits, dim=-1, name=None):
+def softmax(logits, axis=None, name=None, dim=None):
```
You should use the @deprecated_args decorator to mark `dim` deprecated. I am in the process of making that decorator behave less noisy (warn only the first time), but it's already a good idea to use, so usage can adapt to the new names.

yongtang committed:
Use `@deprecated_args`

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@martinwicke Thanks for the review. The PR has been updated using `@deprecated_args`.

drpngx commented:
Jenkins, test this please.

@martinwicke any opinion on the new version?

yongtang commented:
The Jenkins failure is `//tensorflow/tools/api/tests:api_compatibility_test `. Will updated the golden after api changes approved.

asimshankar commented:
(Okay for API change, deferring to @aselle / @martinwicke for the review)

yongtang committed:
Update goldens for API compatibility changes

This commit updates goldens:
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks all for the review. The PR has been rebased and updated with API goldens:
```
    bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
              --update_goldens True
```

Please take a look.

drpngx commented:
Jenkins, test this please.

yongtang commented:
The rest of the tests seems to be fine. For `Linux CPU Tests Makefile` I think it might be unrelated?

drpngx commented:
Jenkins, test this please.

yongtang committed:
Change `dim` to `axis` for `tf.nn.l2_normalize`

to address #14191.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Update goldens for API compatibility changes

This commit updates goldens:
```
bazel-bin/tensorflow/tools/api/tests/api_compatibility_test \
          --update_goldens True
```

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
The PR has been updated to change `dim` to `axis` for `tf.nn.l2_normalize` as well, to address the issue raised in #14191. /cc @girving

martinwicke commented:
Jenkins, test this please.

yongtang commented:
Not sure why the previous Internal build failed as I run it on my dev machine and all passes. I rebased the PR anyway to bring it to the most recent HEAD. I think a rerun of Jenkins will pass.

martinwicke commented:
Jenkins. test this please.

There are currently some issues with our test infrastructure. Sorry about that :./

aselle commented on tensorflow/python/ops/nn_impl.py:
```diff
@@ -333,8 +335,12 @@ def l2_normalize(x, dim, epsilon=1e-12, name=None):
     A `Tensor` with the same shape as `x`.
   """
   with ops.name_scope(name, "l2_normalize", [x]) as name:
+    if dim is not None:
```
could you use deprecation.deprecated_argument_lookup() (it provides this functionality)

yongtang replied:
Thanks @aselle. Updated.

aselle commented on tensorflow/python/ops/nn_ops.py:
```diff
@@ -1643,52 +1644,66 @@ def _swap_axis(logits, dim_index, last_index, name=None):
       dimension of `logits`.
   """
-  return _softmax(logits, gen_nn_ops._softmax, dim, name)
+  if dim is not None:
```
ditto here .

yongtang replied:
Thanks. Done.

aselle reviewed (commented):
thanks for the pr. see comments below.

aselle reviewed (dismissed):
could you use deprecation.deprecated_argument_lookup() (it provides this functionality)

yongtang committed:
Use deprecated_argument_lookup to simplify logic

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @aselle for the review. The PR has been updated with `deprecated_argument_lookup` used. Please take a look and let me know if there are any issues.

asimshankar commented:
(Removing API Review label as the API change is fine)

martinwicke commented:
Jenkins, test this please.

martinwicke commented:
Actually, didn't need testing, changes still to be done.

yongtang commented:
@aselle @martinwicke think the review comment about `deprecated_argument_lookup` has been covered in the update. Please take a look and let me know if there are additional issues.

martinwicke commented:
Oops, sorry, my bad, I didn't see that. LGTM.

martinwicke commented:
I have a suspicion that the failure is due to a broken GPU on gpu6.

Jenkins, test this please.

martinwicke commented:
None of the test failures are related.

girving commented:
Thanks @yongtang!
